src/pam/mod.rs

```diff
@@ -292,25 +274,22 @@ impl<C: Converser> PamContext<C> {
 
     /// Start a user session for the authenticated user.
     pub fn open_session(&mut self) -> PamResult<()> {
-        if !self.session_started {
-            // SAFETY: `self.pamh` contains a correct handle (obtained from `pam_start`).
-            pam_err(unsafe { pam_open_session(self.pamh, self.silent_flag()) })?;
-            self.session_started = true;
-            Ok(())
-        } else {
-            Err(PamError::SessionAlreadyOpen)
-        }
+        assert!(!self.session_started);
```

Comment:
This `assert!` catches my eye (since I think it's the most substantial functional change). 

Is the reasoning here that a `SessionAlreadyOpen` error can only happen due to a error in the use of the PAM module? If so, this comment can be resolved.

Although I would be interested in that case to know if there's a way without `assert!` to force this use case to not happen (e.g. introducing a type that represents and "Opened Session" that consumed the "Unopened Session"), but that doesn't need to go in this PR.

Reply:
Yes, the only places that `session_started` is modified are `open_session` and `close_session`. If we call `open_session` twice without a `close_session` in between that would be a bug.